src/lava/proc/lif/models.py

```diff
@@ -7,29 +7,131 @@
+        self.b_vth_computed = False
+        self.effective_bias = 0
+        self.effective_vth = 0
 
     def run_spk(self):
-        self.u[:] = self.u * ((2 ** 12 - self.du) // 2 ** 12)
+        # Receive synaptic input
         a_in_data = self.a_in.recv()
-        self.u[:] += a_in_data
-        self.v[:] = self.v * \
-            ((2 ** 12 - self.dv) // 2 ** 12) + self.u + self.bias
-        s_out = self.v > self.vth
-        self.v[s_out] = 0  # Reset voltage to 0. This is Loihi-1 compatible.
+
+        # Compute effective bias and threshold only once, not every time-step
+        if not self.b_vth_computed:
+            self.effective_bias = np.left_shift(self.bias, self.bias_exp)
+            # In Loihi, user specified threshold is just the mantissa, with a
+            # constant exponent of 6
+            self.effective_vth = np.left_shift(self.vth, 6)
+            self.b_vth_computed = True
+
+        # Update current
+        # --------------
+        decay_const_u = self.du + self.ds_offset
+        # Below, u is promoted to int64 to avoid overflow of the product
+        # between u and decay constant beyond int32. Subsequent right shift by
+        # 12 brings us back within 24-bits (and hence, within 32-bits)
+        decayed_curr = np.int64(self.u) * (4096 - decay_const_u)
+        decayed_curr = np.sign(decayed_curr) * np.right_shift(np.abs(
+            decayed_curr), 12)
+        decayed_curr = np.int32(decayed_curr)
+        # Hardware left-shifts synpatic input for MSB alignment
+        a_in_data = np.left_shift(a_in_data, 6)
+        # Add synptic input to decayed current
+        decayed_curr += a_in_data
+        # Check if value of current is within bounds of 24-bit. Overflows are
+        # handled by wrapping around modulo 2 ** 23. E.g., (2 ** 23) + k
+        # becomes k and -(2**23 + k) becomes -k
+        wrapped_curr = np.mod(decayed_curr, np.sign(decayed_curr) * (2 ** 23))
+        self.u[:] = wrapped_curr
+        # Update voltage
+        # --------------
+        decay_const_v = self.dv + self.dm_offset
+        # ToDo: make the exponent 23 configurable (see comment above current
+        #  limits)
+        neg_voltage_limit = -np.int32(2 ** 23) + 1
+        pos_voltage_limit = np.int32(2 ** 23) - 1
+        # Decaying voltage similar to current. See the comment above to
+        # understand the need for each of the operations below.
+        decayed_volt = np.int64(self.v) * (4096 - decay_const_v)
+        decayed_volt = np.sign(decayed_volt) * np.right_shift(np.abs(
+            decayed_volt), 12)
+        decayed_volt = np.int32(decayed_volt)
+        updated_volt = decayed_volt + self.u + self.effective_bias
+        self.v[:] = np.clip(updated_volt, neg_voltage_limit, pos_voltage_limit)
+
+        # Spike when exceeds threshold
+        # ----------------------------
+        s_out = self.v >= self.effective_vth
+        # Reset voltage of spiked neurons to 0
+        self.v[s_out] = 0
```

Comment:
Can these magic bit numbers come from descriptive constants instead?

Reply:
Done. Currently defined all magic numbers as instance variables of a ProcessModel instance. We can think of reusing enums from NxCompiler, when that is ready.

Reply:
Perhaps start pulling them out of the LIF class at least into a local ENUM within this module so we find their usage later more easily once we use a more systematic constant database.